### PR TITLE
Feature/gitlab runner module

### PR DESCRIPTION
**Pull request by MaxKavun**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Reviews

Please identify developer to review this change

- [ ] @developer

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass with my changes

### CODE REVIEW

File: tasks/config-gitlab-runner.yml [hunk 1]
```diff
@@ -37,24 +37,19 @@
...
-          {% endfor %}
-          {% if gitlab_runner_extra_options is defined %}
-          {% for extra_option in gitlab_runner_extra_options %}
-          {{ extra_option }}
-          {% endfor %}
-          {% endif %}
-  register: runner_registered
+  gitlab_runner:
+    url: "{{ gitlab_url }}"
+    api_token: ""
```
Comment (tgadiev) [inline tasks/config-gitlab-runner.yml]: Why empty value?

File: tasks/config-gitlab-runner.yml [hunk 2]
```diff
@@ -37,24 +37,19 @@
...
-          --env '{{ env_var }}'
-          {% endfor %}
-          {% if gitlab_runner_extra_options is defined %}
-          {% for extra_option in gitlab_runner_extra_options %}
-          {{ extra_option }}
-          {% endfor %}
-          {% endif %}
-  register: runner_registered
+  gitlab_runner:
+    url: "{{ gitlab_url }}"
```
Comment (tgadiev) [inline tasks/config-gitlab-runner.yml]:
```suggestion
    url: '{{ gitlab_url }}'
```

File: tasks/config-gitlab-runner.yml [hunk 3]
```diff
@@ -37,24 +37,19 @@
...
+    api_token: ""
+    registration_token: '{{ gitlab_ci_token }}'
+    description: '{{ gitlab_runner_description }}'
+    tag_list: '{{ gitlab_runner_tags }}'
+    run_untagged: '{{ gitlab_runner_untagged_builds_run }}'
+    state: present
   become: true
-  failed_when: runner_registered.rc != 0
   when:
     - not gitlab_runner_skip_registration
```
Comment (tgadiev) [inline tasks/config-gitlab-runner.yml]:
```suggestion
    - not gitlab_runner_skip_registration|bool
```

File: tasks/config-gitlab-runner.yml [hunk 4]
```diff
@@ -37,24 +37,19 @@
...
     - gitlab_ci_token | length
     - not (gitlab_runner_description in configured_runners.stderr)
-  environment: '{{ gitlab_runner_config.params }}'
   no_log: "{{ no_logs | default(true) | bool }}"
```
Comment (tgadiev) [inline tasks/config-gitlab-runner.yml]:
```suggestion
  no_log: '{{ no_logs | default(true) | bool }}'
```

File: tasks/prepare-gitlab-runner.yml [hunk 5]
```diff
@@ -17,3 +17,27 @@
...
   tags:
     - always
+
+- name: Install packages on centos
+  yum:
+    name: python-pip
+    state: present
+  retries: 300
+  delay: 20
+  when: ansible_distribution == 'CentOS' or ansible_distribution == 'Red Hat Enterprise Linux'
```
Comment (tgadiev) [inline tasks/prepare-gitlab-runner.yml]: Is it needed to limit this step to Centos or RedHat only?

File: tasks/prepare-gitlab-runner.yml [hunk 6]
```diff
@@ -17,3 +17,27 @@
...
+  register: pip_installed_centos
+  until: pip_installed_centos is succeeded
+  become: true
+  tags:
+    - always
+
+- name: Install Python gitlab module
+  pip:
+    name: python-gitlab==1.12.1
```
Comment (tgadiev) [inline tasks/prepare-gitlab-runner.yml]: Why hardcoding the package version?

**Review by tgadiev — CHANGES_REQUESTED** [review]

File: tasks/prepare-gitlab-runner.yml (hunk 6 above)
Comment (MaxKavun) [reply tasks/prepare-gitlab-runner.yml]: I put version in variable

File: tasks/prepare-gitlab-runner.yml (hunk 5 above)
Comment (MaxKavun) [reply tasks/prepare-gitlab-runner.yml]: Removed red hat from condition

File: tasks/config-gitlab-runner.yml (hunk 1 above)
Comment (MaxKavun) [reply tasks/config-gitlab-runner.yml]: I've put it in variable, near registration_token

File: defaults/main.yml [hunk 7]
```diff
@@ -6,41 +6,40 @@ gitlab_runner_package: >-
     {{ gitlab_version is version('10.0', '>=', strict=true) |
     ternary('gitlab-runner','gitlab-ci-multi-runner') }}
 gitlab_host: git.epam.com
```
Comment (tgadiev) [inline defaults/main.yml]:
```suggestion
gitlab_host: gitlab.com
```

File: tasks/config-gitlab-runner.yml [hunk 8]
```diff
@@ -36,25 +12,32 @@
...
+    api_token: '{{ gitlab_api_token }}'
+    registration_token: '{{ gitlab_ci_token }}'
+    description: '{{ gitlab_runner_description }}'
+    tag_list: '{{ gitlab_runner_tags }}'
+    run_untagged: '{{ gitlab_runner_untagged_builds_run }}'
+    state: present
   become: true
-  failed_when: runner_registered.rc != 0
+  register: gitlab_runner_token
+  changed_when: gitlab_runner_token.changed
```
Comment (tgadiev) [inline tasks/config-gitlab-runner.yml]: ???

File: vars/Debian.yml [hunk 9]
```diff
@@ -4,6 +4,8 @@ packages_base:
   - apt-transport-https
   - debian-archive-keyring
   - ca-certificates
+  - '{{ python_executable }}-pip'
```
Comment (tgadiev) [inline vars/Debian.yml]: If we use `python_executable` variable in role we need to set the default value.

File: vars/RedHat.yml [hunk 10]
```diff
@@ -3,6 +3,7 @@ packages_base:
   - yum-utils
   - pygpgme
   - ca-certificates
+  - python-pip
```
Comment (tgadiev) [inline vars/RedHat.yml]: - '{{ python_executable }}-pip'?

File: tasks/prepare-gitlab-runner.yml [hunk 11]
```diff
@@ -8,6 +8,28 @@
   tags:
     - always
 
+- name: Install EPEL repo.
```
Comment (tgadiev) [inline tasks/prepare-gitlab-runner.yml]: Please, use dynamic include for OS-specific tasks

File: tasks/prepare-gitlab-runner.yml [hunk 12]
```diff
@@ -8,6 +8,28 @@
...
+    state: present
+  when: ansible_distribution == 'CentOS'
+  register: install_epel
+  until: install_epel is succeeded
+  become: true
+  tags:
+    - always
+
+- name: Import EPEL GPG key.
+  rpm_key:
```
Comment (tgadiev) [inline tasks/prepare-gitlab-runner.yml]: Please, use dynamic include for OS-specific tasks

File: tasks/prepare-gitlab-runner.yml [hunk 13]
```diff
@@ -17,3 +39,16 @@
...
+  pip:
+    name:
+      - 'python-gitlab=={{ gitlab_runner_python_module_version }}'
+      - 'python-toml'
+    state: present
+    executable: '{{ pip_executable }}'
+  register: pip_result
+  until: pip_result is succeeded
+  become: true
+  tags:
```
Comment (tgadiev) [inline tasks/prepare-gitlab-runner.yml]: Do we really need tags?

**Review by tgadiev — CHANGES_REQUESTED** [review]

File: tasks/prepare-gitlab-runner.yml [hunk 14]
```diff
@@ -1,19 +1,17 @@
...
   until: installed_packages is succeeded
   become: true
-  tags:
-    - always
+
+- name: Install Python gitlab module
+  pip:
+    name: 'python-gitlab=={{ gitlab_runner_python_module_version }}'
+    state: present
+    executable: '{{ pip_executable | default(pip) }}'
```
Comment (tgadiev) [inline tasks/prepare-gitlab-runner.yml]:
```suggestion
    executable: '{{ pip_executable | default("pip") }}'
```

File: vars/Debian.yml [hunk 15]
```diff
@@ -1,9 +1,16 @@
 ---
+pip_executable: pip3
+python_executable: python3
+ansible_python_interpreter: /usr/bin/python3
```
Comment (tgadiev) [inline vars/Debian.yml]: ansible_* variables are special variables and should not be hardcoded in role.

File: vars/Debian.yml [hunk 16]
```diff
@@ -1,9 +1,16 @@
...
+
 packages_base:
   - gnupg
   - apt-transport-https
   - debian-archive-keyring
   - ca-certificates
+  - '{{ python_executable | default(python3) }}-pip'
```
Comment (tgadiev) [inline vars/Debian.yml]:
```suggestion
  - '{{ python_executable | default("python3") }}-pip'
```

File: tasks/config-gitlab-runner.yml (hunk 8 above)
Comment (MaxKavun) [reply tasks/config-gitlab-runner.yml]: Removed

File: vars/Debian.yml (hunk 9 above)
Comment (MaxKavun) [reply vars/Debian.yml]: Fixed

File: vars/Debian.yml [hunk 17]
```diff
@@ -1,9 +1,16 @@
...
+
+
 gitlab_runner_install_package: "{{ gitlab_runner_package }}\
```
Comment (tgadiev) [inline vars/Debian.yml]:
```suggestion
gitlab_runner_install_package: >-
```

File: vars/RedHat.yml (hunk 10 above)
Comment (MaxKavun) [reply vars/RedHat.yml]: Forgot to add

File: tasks/prepare-gitlab-runner.yml (hunk 11 above)
Comment (MaxKavun) [reply tasks/prepare-gitlab-runner.yml]: Done

File: tasks/prepare-gitlab-runner.yml (hunk 12 above)
Comment (MaxKavun) [reply tasks/prepare-gitlab-runner.yml]: Done

File: tasks/prepare-gitlab-runner.yml (hunk 13 above)
Comment (MaxKavun) [reply tasks/prepare-gitlab-runner.yml]: Someone used them, i've removed them

File: vars/Debian.yml [hunk 18]
```diff
@@ -1,9 +1,16 @@
...
 {{ (gitlab_version != 'latest') | ternary('=' ~ gitlab_version ~ '*', '') }}"
```
Comment (tgadiev) [inline vars/Debian.yml]:
```suggestion
   {{ gitlab_runner_package }}{{ (gitlab_version != 'latest')
       | ternary('=' ~ gitlab_version ~ '*', '') }}
```

File: vars/RedHat.yml [hunk 19]
```diff
@@ -1,8 +1,12 @@
 ---
+pip_executable: pip
+python_executable: python
```
Comment (tgadiev) [inline vars/RedHat.yml]: Hardcode!

File: vars/RedHat.yml [hunk 20]
```diff
@@ -1,8 +1,12 @@
...
+
 packages_base:
   - yum-utils
   - pygpgme
   - ca-certificates
+  - '{{ python_executable | default(python) }}-pip'
```
Comment (tgadiev) [inline vars/RedHat.yml]:
```suggestion
  - '{{ python_executable | default("python") }}-pip'
```

**Review by tgadiev — CHANGES_REQUESTED** [review]

File: vars/Debian.yml (hunk 15 above)
Comment (MaxKavun) [reply vars/Debian.yml]: fixed

File: vars/RedHat.yml (hunk 19 above)
Comment (MaxKavun) [reply vars/RedHat.yml]: Replaced

File: tasks/config-gitlab-runner.yml [hunk 21]
```diff
@@ -1,60 +1,32 @@
...
+    registration_token: '{{ gitlab_ci_token }}'
+    description: '{{ gitlab_runner_description }}'
+    tag_list: '{{ gitlab_runner_tags }}'
+    run_untagged: '{{ gitlab_runner_untagged_builds_run }}'
+    state: present
   become: true
-  failed_when: runner_registered.rc != 0
+  register: gitlab_runner_token
   when:
     - not gitlab_runner_skip_registration
```
Comment (tgadiev) [inline tasks/config-gitlab-runner.yml]:
```suggestion
    - not gitlab_runner_skip_registration|bool
```

File: vars/RedHat.yml [hunk 22]
```diff
@@ -1,11 +1,14 @@
...
+gitlab_runner_packages_base:
   - yum-utils
-  - pygpgme
   - ca-certificates
+  - '{{ python_executable | default("python") }}-pip'
 
 gitlab_runner_install_package: "{{ gitlab_runner_package }}\
 {{ (gitlab_version != 'latest') | ternary('-' ~ gitlab_version ~ '*', '') }}"
 
 gitlab_runner_repo: "https://packages.gitlab.com/runner/\
```
Comment (tgadiev) [inline vars/RedHat.yml]:
```suggestion
gitlab_runner_repo: >-
```

File: vars/RedHat.yml [hunk 23]
```diff
@@ -1,11 +1,14 @@
...
-{{ gitlab_runner_package }}/el/{{ ansible_distribution_major_version }}/$basearch"
+{{ gitlab_runner_package }}/el/7/$basearch"
```
Comment (tgadiev) [inline vars/RedHat.yml]:
```suggestion
    https://packages.gitlab.com/runner/{{ gitlab_runner_package }}/el/7/$basearch
```

**Review by tgadiev — CHANGES_REQUESTED** [review]

File: defaults/main.yml [hunk 24]
```diff
@@ -1,48 +1,52 @@
...
 # defaults file for ansible-role-gitlab-runner
-
-gitlab_version: '10.5'
+yum_libselinux_python_library: libselinux-python
+yum_libselinux_config_libraries:
+  - policycoreutils-python
+  - libsemanage-python
+epel_repository_url: 'https://dl.fedoraproject.org/pub/epel/epel-release-latest-{{ ansible_distribution_major_version }}.noarch.rpm'
+epel_rpm_key: '/etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-{{ ansible_distribution_major_version }}'
+gitlab_version: '12.4'
```
Comment (tgadiev) [inline defaults/main.yml]:
```suggestion
gitlab_version: latest
```

**Review by tgadiev — CHANGES_REQUESTED** [review]

**Review by pavelpikta — DISMISSED** [review]

**Review by tgadiev — APPROVED** [review]

**Review by pavelpikta — APPROVED** [review]

**Pull request merged**